extended information for slice pattern errors (E0527 through E0529)

zackmdavis opened the pull request:
r? @GuillaumeGomez

GuillaumeGomez commented on src/librustc_typeck/diagnostics.rs:
````diff
@@ -4073,6 +4073,111 @@ impl SpaceLlama for i32 {
+Example of erroneous code:
+
+```compile_fail
+#![feature(slice_patterns)]
````
This code throws E0308 on nightly as it seems.

GuillaumeGomez replied:
Now it does.

GuillaumeGomez commented on src/librustc_typeck/diagnostics.rs:
````diff
@@ -4073,6 +4073,111 @@ impl SpaceLlama for i32 {
+Example of erroneous code:
+
+```compile_fail
+#![feature(slice_patterns)]
````
This code throws E0308 as well.

GuillaumeGomez commented on src/librustc_typeck/diagnostics.rs:
````diff
@@ -4073,6 +4073,111 @@ impl SpaceLlama for i32 {
+Example of erroneous code:
+
+```compile_fail
+#![feature(slice_patterns)]
````
And this one throws E0308 as well.

GuillaumeGomez commented:
Have you tested your code on nightly?

zackmdavis commented:
That's strange! What specific nightly are you on? I composed these against 5c2a5d449 2016-06-11, and it looks like they still behave the same against bb4a79b08 2016-06-15 (which `multirust update nightly` just gave me right now):

```
zmd@ExpectedReturn:~/Code/Misc$ rustc --version
rustc 1.11.0-nightly (bb4a79b08 2016-06-15)
zmd@ExpectedReturn:~/Code/Misc$ cat e0527.rs 
#![feature(slice_patterns)]

fn main() {
    let r = &[1, 2, 3, 4];
    match r {
        &[a, b] => { // error: pattern requires 2 elements but array
                     //        has 4
            println!("a={}, b={}", a, b);
        }
    }
}
zmd@ExpectedReturn:~/Code/Misc$ rustc e0527.rs 
e0527.rs:6:10: 6:16 error: pattern requires 2 elements but array has 4 [E0527]
e0527.rs:6         &[a, b] => { // error: pattern requires 2 elements but array
                    ^~~~~~
error: aborting due to previous error
zmd@ExpectedReturn:~/Code/Misc$ cat e0528.rs 
#![feature(slice_patterns)]

fn main() {
    let r = &[1, 2];
    match r {
        &[a, b, c, rest..] => { // error: pattern requires at least 3
                                //        elements but array has 2
            println!("a={}, b={}, c={} rest={:?}", a, b, c, rest);
        }
    }
}
zmd@ExpectedReturn:~/Code/Misc$ rustc e0528.rs 
e0528.rs:6:10: 6:27 error: pattern requires at least 3 elements but array has 2 [E0528]
e0528.rs:6         &[a, b, c, rest..] => { // error: pattern requires at least 3
                    ^~~~~~~~~~~~~~~~~
error: aborting due to previous error
zmd@ExpectedReturn:~/Code/Misc$ cat e0529.rs 
#![feature(slice_patterns)]

fn main() {
    let r: f32 = 1.0;
    match r {
        [a, b] => { // error: expected an array or slice, found `f32`
            println!("a={}, b={}", a, b);
        }
    }
}
zmd@ExpectedReturn:~/Code/Misc$ rustc e0529.rs 
e0529.rs:6:9: 6:15 error: expected an array or slice, found `f32` [E0529]
e0529.rs:6         [a, b] => { // error: expected an array or slice, found `f32`
                   ^~~~~~
error: aborting due to previous error
```

zackmdavis commented:
Also, Travis points out that I made a syntax error (which I suppose is only just punishment for relying on CI instead of building it locally); I'll (force-)push a fix for that later.

GuillaumeGomez commented:
I tried on play.rust-lang.org. So maybe the nightly version isn't up-to-date. Don't really know. I'll retry you'll update it.

zackmdavis commented:
@GuillaumeGomez Force-pushed to fix syntax. The playpen was very out of date, but has been fixed now (rust-lang/rust-playpen#215).

zackmdavis commented:
Force-pushed again because the corrected code samples were failing tests due to the lack of the `#![feature(slice_patterns)]` attribute. I've marked them as `ignore` because I had deliberately left the attribute out because it's already in the erroneous code samples just above and I want to conserve vertical terminal space for people running `--explain`.

GuillaumeGomez commented on src/librustc_typeck/diagnostics.rs:
````diff
@@ -4073,6 +4073,110 @@ impl SpaceLlama for i32 {
+
+Example of erroneous code:
+
+```compile_fail
````
Please add "E0527 flag", like this:

```
```compile_fail,E0527
```

GuillaumeGomez commented on src/librustc_typeck/diagnostics.rs:
````diff
@@ -4073,6 +4073,110 @@ impl SpaceLlama for i32 {
+
+Example of erroneous code:
+
+```compile_fail
````
Please add "E0528 flag", like this:

```
```compile_fail,E0528 
```

GuillaumeGomez commented on src/librustc_typeck/diagnostics.rs:
````diff
@@ -4073,6 +4073,110 @@ impl SpaceLlama for i32 {
+
+Example of erroneous code:
+
+```compile_fail
````
Please add "E0529 flag", like this:

```
```compile_fail,E0529 
```

GuillaumeGomez commented:
About the error code flags, take a look [here](https://github.com/rust-lang/rust/pull/34186) for more information.

zackmdavis commented:
(added error code flags)

GuillaumeGomez commented:
Everything looks good. Can you make a commit for each error code explanation please?

zackmdavis commented:
@GuillaumeGomez Force-pushed with error code test directives squashed into the individual per-extended-error commits (and rebased on master). I regret the delay.

GuillaumeGomez commented on src/librustc_typeck/diagnostics.rs:
````diff
@@ -4092,6 +4092,110 @@ impl SpaceLlama for i32 {
+```compile_fail,E0527
+#![feature(slice_patterns)]
+
+fn main() {
````
No need for main function.

GuillaumeGomez commented on src/librustc_typeck/diagnostics.rs:
````diff
@@ -4092,6 +4092,110 @@ impl SpaceLlama for i32 {
+array. Additional elements can be matched with `..`:
+
+```ignore
+fn main() {
````
No need for main function.

GuillaumeGomez commented on src/librustc_typeck/diagnostics.rs:
````diff
@@ -4092,6 +4092,110 @@ impl SpaceLlama for i32 {
+Ensure that the pattern is consistent with the size of the matched
+array. Additional elements can be matched with `..`:
+
+```ignore
````
Why adding ignore flag?

zackmdavis replied:
See my [earlier comment](https://github.com/rust-lang/rust/pull/34319#issuecomment-227732647):

> [...] the corrected code samples were failing tests due to the lack of the `#![feature(slice_patterns)]` attribute. I've marked them as `ignore` because I had deliberately left the attribute out because it's already in the erroneous code samples just above and I want to conserve vertical terminal space for people running `--explain`.

GuillaumeGomez replied:
Then please add it. :)

It's part of the error after all.

GuillaumeGomez commented on src/librustc_typeck/diagnostics.rs:
````diff
@@ -4092,6 +4092,110 @@ impl SpaceLlama for i32 {
+```compile_fail,E0528
+#![feature(slice_patterns)]
+
+fn main() {
````
No need for main function.

GuillaumeGomez commented on src/librustc_typeck/diagnostics.rs:
````diff
@@ -4092,6 +4092,110 @@ impl SpaceLlama for i32 {
+non-`..` elements in the pattern:
+
+```ignore
+fn main() {
````
No need for main function.

GuillaumeGomez commented on src/librustc_typeck/diagnostics.rs:
````diff
@@ -4092,6 +4092,110 @@ impl SpaceLlama for i32 {
+Ensure that the matched array has at least as many elements as there are
+non-`..` elements in the pattern:
+
+```ignore
````
Why adding ignore flag?

GuillaumeGomez commented on src/librustc_typeck/diagnostics.rs:
````diff
@@ -4092,6 +4092,110 @@ impl SpaceLlama for i32 {
+}
+```
+
+Ensure that the matched array has at least as many elements as there are
````
This isn't very explicit. What about: "Ensure that the matched array has the same number of elements. If you don't need to bind them all, just add `..` or `variable..`"?

GuillaumeGomez commented on src/librustc_typeck/diagnostics.rs:
````diff
@@ -4092,6 +4092,110 @@ impl SpaceLlama for i32 {
+```compile_fail,E0529
+#![feature(slice_patterns)]
+
+fn main() {
````
No need for main function.

GuillaumeGomez commented on src/librustc_typeck/diagnostics.rs:
````diff
@@ -4092,6 +4092,110 @@ impl SpaceLlama for i32 {
+types:
+
+```ignore
+fn main() {
````
No need for main function.

GuillaumeGomez commented on src/librustc_typeck/diagnostics.rs:
````diff
@@ -4092,6 +4092,110 @@ impl SpaceLlama for i32 {
+Ensure that the pattern and the expression being matched on are of consistent
+types:
+
+```ignore
````
Why the ignore flag?

GuillaumeGomez commented:
No problem. A few nits to fix and it should be good to go!

bors commented:
:umbrella: The latest upstream changes (presumably #34365) made this pull request unmergeable. Please resolve the merge conflicts.

zackmdavis commented:
(force-pushed)

GuillaumeGomez commented on src/librustc_typeck/diagnostics.rs:
````diff
@@ -4029,6 +4029,98 @@ impl SpaceLlama for i32 {
+Ensure that the matched array has as many elements as the pattern
+requires. Additional elements can be matched with `..`.
+
+```ignore
````
Why the ignore flag?

bors commented:
:umbrella: The latest upstream changes (presumably #34907) made this pull request unmergeable. Please resolve the merge conflicts.

zackmdavis commented:
(force-pushed again)

GuillaumeGomez commented:
All good except the last [remaining nit](https://github.com/rust-lang/rust/pull/34319#issuecomment-229604771).

zackmdavis commented:
Oh, right; in the abyssal depths of my carelessness, I removed the ultimately unneeded `ignore` annotations and `main()` functions but forgot to [rephrase the sentence](https://github.com/rust-lang/rust/pull/34319#discussion_r69098364) explaining how to fix E0528. What if we split it into two sentences, like:

```
Ensure that the matched array has at least as many elements as the pattern
requires. You can match an arbitrary number of remaining elements with `..`:
```
?

(forced-pushed again)

GuillaumeGomez commented:
Seems good for me! Waiting for your update then!

zackmdavis commented:
To clarify, e960021 already has the revised sentences; this may be ready to merge.

GuillaumeGomez commented:
Sorry, didn't check... Thanks for your work!

@bors: r+ rollup

bors commented:
:pushpin: Commit e960021 has been approved by `GuillaumeGomez`

<!-- @bors r=GuillaumeGomez e960021d30faff871f98c9c463b43c4359af280d -->

bors commented:
:hourglass: Testing commit e960021 with merge c12c37a...

alexcrichton commented:
@bors: retry force clean

* restarted buildbot

bors commented:
:bomb: Buildbot returned an error: `<span>exceptions.KeyError</span>: <span>'auto-linux-64-opt-mir'</span>`

bors commented:
:hourglass: Testing commit e960021 with merge 505842a...

bors commented:
:broken_heart: Test failed - [auto-win-msvc-64-opt-no-mir](https://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt-no-mir/builds/0)

alexcrichton commented:
@bors: retry

* https://github.com/rust-lang/rust/pull/35187

On Tue, Aug 2, 2016 at 12:05 PM, bors <notifications@github.com> wrote:

> 💔 Test failed - auto-win-msvc-64-opt-no-mir
> <https://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt-no-mir/builds/0>
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/rust-lang/rust/pull/34319#issuecomment-237009395>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAD95OKkT2PEB1Ajs3mPnwK5fEXSFr3Cks5qb5T0gaJpZM4I4Fy2>
> .
>
